craft_providers/executor.py

```diff
@@ -0,0 +1,101 @@
+# Copyright (C) 2021 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Executor module."""
+import logging
+import pathlib
+import subprocess
+from abc import ABC, abstractmethod
+from typing import List
+
+logger = logging.getLogger(__name__)
+
+
+class Executor(ABC):
+    """Interfaces to execute commands and move data in/out of an environment."""
+
+    @abstractmethod
+    def create_file(
+        self,
+        *,
+        destination: pathlib.Path,
+        content: bytes,
+        file_mode: str,
+        group: str = "root",
+        user: str = "root",
+    ) -> None:
+        """Create file with content and file mode.
+
+        :param destination: Path to file.
+        :param content: Contents of file.
+        :param file_mode: File mode string (e.g. '0644').
+        :param group: File owner group ID.
+        :param user: Filer owner user ID.
```

Comment:
This contradicts the default above... is it the user/group *id* or the *name*?

Reply:
It could be either, as supported by chown.  Will update docs.

Reply:
Fixed